CXX-3309 Automate SBOM generation and Endor Labs scanning

jasonhills-mongodb opened the pull request:
### Objective
- Obviate manual SBOM component management via automated generation of a CycloneDX SBOM that includes all required and optional runtime components.
- Add Endor Labs SCA scanning of PRs (non-blocking)

### Changes
- Add a GitHub Action workflow (`.github/workflows/endor_scan_and_generate_sbom.yml`) triggered by edits to cmake files that:
  - (`push`) performs an Endor Labs SCA scan and exports an SBOM for `master` or `releases/v*` branches, enriches SBOM with metadata, opens a PR for updated SBOM
  - (`pull_request`) performs an Endor Labs SCA scan for PRs (non-blocking) and, if potential vulnerabilities were found, add a comment to PR 
  - (`workflow_dispatch`) workflow can also be triggered manually
- Remove the `etc/purls.txt` file
  - update all references to it in scripts and documentation
  - existing scripts and processes using Silkbomb for the augmented SBOM are unchanged, except that the `sbom.json` file is used as input instead of `etc/purls.txt`
- Change the location of the SBOM file from `etc/cyclonedx.sbom.json` to `sbom.json` for consistency with other MDB repos
- Add `etc/sbom` folder with:
  - Python scripts for SBOM generation using Endor Labs scan results and pre-defined enrichment data
    - Dependency group added to `pyproject.toml`
  - SBOM enrichment data in `etc/sbom/metadata.cdx.json`
- Updated SBOM-related documentation in `etc/releasing.md`


### Testing
The workflow was thoroughly tested on a fork, including scenarios with missing or malformed files.

### Permissions
The workflow requires the repository configuration to allow it to request write access and open PRs.

### Miscellaneous
- Once this is merged to `master`, it may be cherry picked to the relevant release branches.
- Once approved, this will be ported to the C driver repo.
- The new SBOM format is in alignment with the [NITA Minimum Elements for Software Bill of Materials](https://www.ntia.gov/files/ntia/publications/sbom_minimum_elements_report.pdf) and OWASP [Software Component Verification Standard (SCVS)](https://scvs.owasp.org/scvs/). Both publications emphasize that all components used in creation of software are to be documented in an SBOM, including ecosystem-based dependencies, as a best practice.

jasonhills-mongodb committed:
CXX-3309 Automate SBOM generation using Endor Labs

eramongodb commented on CMakeLists.txt:
```diff
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+
```
```suggestion
```
Stray whitespace?

jasonhills-mongodb committed:
Add updated SBOM

eramongodb commented on .github/workflows/endor_scan_and_generate_sbom.yml:
```diff
@@ -0,0 +1,144 @@
+        working-directory: ${{env.BUILD}}
+        run: |
+          cmake .. -DCMAKE_BUILD_TYPE=${{env.BUILD_TYPE}} -DCMAKE_CXX_STANDARD=${{env.CXX_STANDARD}} -DENABLE_TESTS=ON
+          rm .gitignore # prevent exclusion of build/_deps from endorctl scan
```
```suggestion
          echo "!_deps" >>.gitignore # prevent exclusion of build/_deps from endorctl scan
```
Consider excluding the `_deps` directory specifically instead.

eramongodb commented on .github/workflows/endor_scan_and_generate_sbom.yml:
```diff
@@ -0,0 +1,144 @@
+        env:
+          ENDOR_SCAN_EMBEDDINGS: true
+        with:
+          additional_args: '--languages=c --include-path="build/_deps/**"'
```
Should the `--languages` flag use `c,cpp` instead? Catch2 (only used for tests) is a C++ library.

jasonhills-mongodb replied:
In endorctl, you indicate "c" to enable both C and C++

jasonhills-mongodb commented on .github/scripts/sbom/config.py:
Good call. Will do.

jasonhills-mongodb committed:
CXX-3309 Refactor Endor Labs SBOM workflow and sync scripts

- Split monolithic endor_scan_and_generate_sbom.yml into separate
  endor_monitoring_sbom.yml (push/dispatch) and endor_pr_scan.yml
  (pull_request) workflows
- Extract shared clone+CMake configure steps into composite action
  .github/actions/clone-and-configure
- Move SBOM scripts from etc/sbom/ to .github/scripts/sbom/
- Sync config.py, generate_sbom.py, endorctl_utils.py from mongodb/mongo,
  removing mongo-specific components and applying branch-context bug fix
- Replace Kondukto token auth with AWS credential pass-through in
  Evergreen sbom.sh and config generator
- Update CODEOWNERS for new workflow and action paths

bisht2050 commented:
@jasonhills-mongodb Do you still intend to work on this and get it eventually merged?

kevinAlbs commented:
@jasonhills-mongodb is this OK to close? Or do you intend to continue on this PR?

kevinAlbs commented:
Closing due to staleness. Feel free to re-open or open a new PR.
